### ISSUE TITLE
Race condition when determining provider for child module data block

### ISSUE THREAD

**Issue by talton-heb** [hashicorp/terraform#21699]
### Terraform Version
```
$ terraform -v
Terraform v0.12.1
+ provider.aws v2.14.0
```

### Terraform Configuration Files

`main.tf`
```hcl
provider "aws" {
  version = "2.14.0"
  region = "us-east-1"
}
data "aws_region" "current" {}
output "region" {
  value = data.aws_region.current
}
module "child" {
  source = "./child"
  providers = {
    aws = aws
  }
}
output "child_region" {
  value = module.child.region
}
```

`child/main.tf`
```hcl
provider "aws" {
  version = "2.14.0"
  region = "us-west-1"
}
data "aws_region" "current" {}
output "region" {
  value = data.aws_region.current
}

```

### Debug Output

[Log from when `region` and `child_region` outputs are the same](https://gist.github.com/talton-heb/994c89d39eea48b8e2c7dfe04696dc39)

[Log from when `region` and `child_region` outputs are different](https://gist.github.com/talton-heb/3dfbff2e66346c25b901a3283ef2343b)

### Expected Behavior
The `data` block in `child/main.tf` module should have inherited the region from the provider explicitly configured in the `module` block in `main.tf`.

Expected out:
```
Outputs:

child_region = {
  "description" = "US East (N. Virginia)"
  "endpoint" = "ec2.us-east-1.amazonaws.com"
  "id" = "us-east-1"
  "name" = "us-east-1"
}
region = {
  "description" = "US East (N. Virginia)"
  "endpoint" = "ec2.us-east-1.amazonaws.com"
  "id" = "us-east-1"
  "name" = "us-east-1"
}
```

### Actual Behavior
Sometimes it outputs:
```
Outputs:

child_region = {
  "description" = "US East (N. Virginia)"
  "endpoint" = "ec2.us-east-1.amazonaws.com"
  "id" = "us-east-1"
  "name" = "us-east-1"
}
region = {
  "description" = "US East (N. Virginia)"
  "endpoint" = "ec2.us-east-1.amazonaws.com"
  "id" = "us-east-1"
  "name" = "us-east-1"
}
```

and sometimes it outputs:
```
Outputs:

child_region = {
  "description" = "US West (N. California)"
  "endpoint" = "ec2.us-west-1.amazonaws.com"
  "id" = "us-west-1"
  "name" = "us-west-1"
}
region = {
  "description" = "US East (N. Virginia)"
  "endpoint" = "ec2.us-east-1.amazonaws.com"
  "id" = "us-east-1"
  "name" = "us-east-1"
}
```

### Steps to Reproduce
Keep running `terraform apply` until the output changes.

### Additional Context
Using [`aws-vault`](https://github.com/99designs/aws-vault) to supply AWS creds via environment variables. 

### References
None

**Comment by jbardin** [comment]
Thanks for filing the issue @talton-heb! 

Your example with the complete logs (each log actually demonstrates the inconsistency between the different phases) should give us what we need to track this down. The resource in the module is somehow being connected with an arbitrary "aws" provider, rather than the one within the module. 

Your config actually should _not_ inherit the provider from the top level, because a provider "proxy configuration" cannot contain anything other than an alias. (https://www.terraform.io/docs/configuration/modules.html#passing-providers-explicitly)
This is creating a separate provider within the module, though it is only being connected in the graph half the time. You can't override providers configured in modules, which is partly why we recommend only configuring them at the top level. In order to specify provider versions globally, you can also use the `required_providers` argument in the `"terraform"` configuration block: https://www.terraform.io/docs/configuration/terraform.html#specifying-required-provider-versions

**Comment by talton-heb** [comment]
Hey @jbardin!  Thanks for looking at this issue.

#### Question 1
Following your "Passing Providers Explicitly" link, there does not seem to be any wording that "proxy configuration" should only contain the `alias` attribute.  Should I open another issue for adding that to the documentation?

#### Question 2
I'm working on an in-house library of Terraform modules, and as part of the transition to 0.12 I'm trying to understand how to properly write the modules in the library to support both (a) providing the `provider` configuration from the root module and (b) passing `terraform validate` successfully.  When I run validate against `child/main.tf` (like I in my original post), validation passes:
```bash
[10:25:08] ~/Work/experiments/nested_providers
$ terraform validate child/
Success! The configuration is **valid.**
```

However, if I modify the provider in `child/main.tf` to remove all properties except for alias as suggested by your comment:

```bash
[10:28:35] ~/Work/experiments/nested_providers
$ cat child/main.tf
provider "aws" {
  alias = "child"
}
data "aws_region" "current" {}
output "region" {
  value = data.aws_region.current
}
[10:28:54] ~/Work/experiments/nested_providers
$ terraform validate child

Error: Missing required argument

The argument "region" is required, but was not set.
```

This seems to be inconsistent.  Should I open another issue against the `validate` command?

#### Question 3
When a module in our library is updated to take advantage of a feature in a new version of a provider (eg: [CodeBuild Webhoook FilterGroups](https://github.com/terraform-providers/terraform-provider-aws/pull/8110) in AWS 2.14), it would be handy to be able to express this requirement in the `provider` block inside the library module. This seems to be supported by setting the `version` inside both parent and child module `provider` blocks.  

For example, if we make the following edits to `child/main.tf`:
```hcl
provider "aws" {
  version = "~> 2.14.0"
}
```
and `main.tf`:
```hcl
provider "aws" {
  version = "2.13.0"
}
```
and try to init or apply it, we get the following:
```bash
[10:40:13] ~/Work/experiments/nested_providers
$ terraform apply

Error: Could not satisfy plugin requirements


Plugin reinitialization required. Please run "terraform init".

Plugins are external binaries that Terraform uses to access and manipulate
resources. The configuration provided requires plugins which can't be located,
don't satisfy the version constraints, or are otherwise incompatible.

Terraform automatically discovers provider requirements from your
configuration, including providers used in child modules. To see the
requirements and constraints from each module, run "terraform providers".



Error: provider.aws: no suitable version installed
  version requirements: "2.13.0,~> 2.14.0"
  versions installed: "2.14.0"


[10:40:17] ~/Work/experiments/nested_providers
$ terraform init
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...

No provider "aws" plugins meet the constraint "2.13.0,~> 2.14.0".

The version constraint is derived from the "version" argument within the
provider "aws" block in configuration. Child modules may also apply
provider version constraints. To view the provider versions requested by each
module in the current configuration, run "terraform providers".

To proceed, the version constraints for this provider must be relaxed by
either adjusting or removing the "version" argument in the provider blocks
throughout the configuration.


Error: no suitable version is available
```

Can you clarify if your statement also means the proxy config _cannot_ contain version information?  If the proxy config should not contain version information, what is the correct way for a library module to express it's requirements?


Thanks again for your time!

**Comment by jbardin** [comment]
Hi @talton-heb,

Q1: Oh yes, that probably should be mentioned somewhere in that section. It was probably lost during revision.

Q2: Please file an issue for this. I'm not sure how we want to document validating modules independently, but I can see how it could lead you to assuming you need a complete config for the providers.

Q3: Terraform only supports a single provider version for a given config, so there must be an available version that matches the logical AND of all the constraints throughout the config. Due to some existing internal behavior, the `version` argument for the provider also prevents it from being used as a "proxy" placeholder within a module, but since only one version is allowed anyway, we decided to add the global `required_providers` argument so it can more easily be defined in one place. 

Any ideas to help make this easier to find in the documentation are welcome!

**Comment by apparentlymart** [comment]
The problem with running `terraform validate` directly on a child module is, I think, the same as what's being discussed in #21408.

**Comment by talton-heb** [comment]
@apparentlymart - After reading through the conversation, I concur that #21706 duplicates #21408 (it is perhaps a bit more succinct, though).  Would you like me to close it as a dupe?

**Comment by mwaaas** [comment]
experiencing the same issue with provider alias. 
in the child module I have 

terraform {
  required_version = ">= 0.12"
  required_providers {
    aws = "~> 2.34"
  }
}

provider "aws" {
  region = var.region
}

then my  resources 

in my main module I have 

provider "aws" {
  region  = "eu-west-1"
  version = "~> 2.31"
  alias = "default_customer_profile_provider"
  profile = var.AWS_PROFILE
  endpoints {
    dynamodb = var.AWS_DYNAMODB_ENDPOINT
    sqs      = var.AWS_SQS_ENDPOINT
    sns      = var.AWS_SNS_ENDPOINT
    sts      = var.AWS_STS_ENDPOINT
  }
  assume_role {
    role_arn = var.TERRAFORM_ROLE_ARN
  }
}

module "customer_profile_table" {
  source         = "child_module"
  providers = {
    aws = "aws.default_customer_profile_provider"
  }
}


Sometimes it picks endpoints for child sometimes it picks endpoint for my main module

**Comment by mwaaas** [comment]
I have managed to reproduce the issue 
https://github.com/mwaaas/reproduce_bug_in_terraform

Any pointers on how I can contribute to this

**Comment by jbardin** [comment]
Hi @mwaaas,

Thanks for the additional data point. The issue is well understood, so I don't think there's anything else that needs to be contributed. Your example isn't passing a provider into the module, because the module `aws` config in the module isn't empty and creates a separate provider instance, in which case Terraform is not resolving the ambiguity correctly. The correct behavior for this configuration is to not use `aws.default_customer_profile_provider`, so fixing the configuration should resolve the issue for you.

**Comment by mwaaas** [comment]
Hi @jbardin 

thanks for the clarification, 

So if I remove the provider from the module it will work correctly, but was trying to simulate a scenario where I don't own the module, let's say am importing it from Github and I want to override the provider 

ie. 
module "event_table" {
  source         = "githbu/tf_modules/dyanamodb"

  providers = {
    aws = aws.custom_provider
  }
}

**Comment by jbardin** [comment]
@mwaaas unfortunately that's not the direction that was chosen for overrides. You can't override a provider in a module, the module can only override a provider from a parent scope. In general, shared modules should not define their providers so that they can all be configured externally.

**Comment by mwaaas** [comment]
@jbardin thanks didn't know that. 

What about this behaviour of sometimes it overrides sometimes it does not, should that be fixed. 
What you had mentioned earlier as _**Terraform is not resolving the ambiguity correctly**_.

**Comment by jbardin** [comment]
Hi All,

This type of issue should be resolved with the 0.15 release. Along with various bug fixes, provider configurations for modules are more strictly validated, and more contextual diagnostics are emitted to help guide users to a supported configuration. Any similar problems with current releases of Terraform should be filed in new issues.

Thanks!
